internal/internal.go

```diff
@@ -345,25 +353,34 @@ func (p *envoyExtAuthzGrpcServer) Check(ctx context.Context, req *ext_authz_v3.C
 	if code := stop(); resp != nil && code != nil {
 		resp.Status = code
 	}
-	return resp, err
+
+	if err != nil {
+		return resp, err.Unwrap()
+	}
+	return resp, nil
 }
 
-func (p *envoyExtAuthzGrpcServer) check(ctx context.Context, req interface{}) (*ext_authz_v3.CheckResponse, func() *rpc_status.Status, error) {
+func (p *envoyExtAuthzGrpcServer) check(ctx context.Context, req interface{}) (*ext_authz_v3.CheckResponse, func() *rpc_status.Status, *Error) {
 	var err error
 	var evalErr error
+	var internalErr Error
 	start := time.Now()
 	logger := p.manager.Logger()
 
 	result, stopeval, err := envoyauth.NewEvalResult()
 	if err != nil {
 		logger.WithFields(map[string]interface{}{"err": err}).Error("Unable to start new evaluation.")
-		return nil, func() *rpc_status.Status { return nil }, err
+		internalErr = Error{Code: StartCheckErr}
+		internalErr.Wrap(err)
+		return nil, func() *rpc_status.Status { return nil }, &internalErr
 	}
 
 	txn, txnClose, err := result.GetTxn(ctx, p.Store())
 	if err != nil {
 		logger.WithFields(map[string]interface{}{"err": err}).Error("Unable to start new storage transaction.")
-		return nil, func() *rpc_status.Status { return nil }, err
+		internalErr = Error{Code: StartTxnErr}
+		internalErr.Wrap(err)
```

Comment:
Nit: Since this is repeated in many places, might be better to pull into a separate routine.